src/main.rs

```diff
@@ -357,55 +355,47 @@ fn build_config(version: &str) -> Config {
-fn provision(config: &Config) -> () {
-    let tls  = config.tls.as_ref().expect("tls config required");
-
-    if !Path::new(&tls.p12_path).exists() {
-        let prov = config.provision.as_ref().expect("provision config required");
-        init_tls_client_from_p12(&prov.p12_path, &prov.p12_password);
-        let url = tls.server.join("/devices");
-        let id  = prov.device_id.as_ref().unwrap_or(&config.device.uuid);
-        let http = AuthClient::default();
-
-        loop {
-            match pkcs12(url.clone(), id.clone(), prov.expiry_days, &http) {
-                Ok(bundle) => {
-                    let _ = File::create(&tls.p12_path)
-                        .map(|mut file| file.write(&*bundle).expect("couldn't write pkcs12 bundle"))
-                        .map_err(|err| panic!("couldn't open tls.p12_path for writing: {}", err));
-                    break;
-                }
-                Err(e) => {
-                    // delay of 0 means no retry
-                    if prov.retry_delay_seconds == 0 {
-                        panic!("Error while getting bundle: {}", e);
-                    } else {
-                        info!("Retrying after error while getting bundle: {}", e);
-                        thread::sleep(Duration::from_secs(prov.retry_delay_seconds));
-                    }
-                }
-            }
-        }
+/// Initialize the client then return the interpreter processing mode.
+fn initialize(config: &Config, auth: &Auth) -> CommandMode {
+    if *auth == Auth::Provision {
+        provision_p12(&config);
     }
-}
 
-fn init_config(config: &Config) -> CommandMode {
-    if let Some(ref tls) = config.tls {
-        if let Some(ref certificates_path) = tls.certificates_path {
-            init_tls_client(TlsData { ca_path: Some(certificates_path), p12_path: Some(&tls.p12_path),
-                                      p12_pass: Some(&tls.p12_password) });
-        } else {
-            init_tls_client_from_p12(&tls.p12_path, &tls.p12_password);
-        }
-    } else {
-        let ca_path = config.device.certificates_path.as_ref().map(Deref::deref);
-        init_tls_client(TlsData { ca_path: ca_path, p12_path: None, p12_pass: None });
-    }
+    init_tls_client(config.tls_data());
 
     if let PackageManager::Uptane = config.device.package_manager {
-        CommandMode::Uptane(None)
+        let uptane = Uptane::new(config).unwrap_or_else(|err| exit!(2, "couldn't start uptane: {}", err));
+        CommandMode::Uptane(uptane)
     } else {
         CommandMode::Sota
     }
 }
+
+/// Download the PKCS#12 bundle needed to communicate with the server.
+fn provision_p12(config: &Config) -> () {
+    let prov = config.provision.as_ref().expect("provisioning expects a [provision] config");
+    let tls  = config.tls.as_ref().expect("provisioning expects a [tls] config");
+
+    if Path::new(&tls.p12_path).exists() {
+        exit!(3, "{}", "can't provision when tls.p12_path already exists");
```

Comment:
Is it possible we created the file, but didn't provision correctly. How would this be error handled?